### PR TITLE
Change map helper functions' arguments

### LINKED ISSUE #451: Use map_from_arrays() for map construction in rustler_codegen
As noted in https://github.com/rusterlium/rustler/pull/440, we should try to use `map_from_arrays()` when constructing maps in `rustler_codegen`. As pointed out by @SeokminHong, changing the signature of `map_from_arrays()` to use `impl Encoder` instead of `Term<'a>` could simplify that:

```rust
    pub fn map_from_arrays(
        env: Env<'a>,
        keys: &[impl Encoder],
        values: &[impl Encoder],
    ) -> NifResult<Term<'a>> {
```

### PR DESCRIPTION
**Pull request by SeokminHong**
resolves #451

Change map's helper functions arguments from `Term` into `impl Encoder`.

### CODE REVIEW

**Review by philss — APPROVED** [review]
LGTM! :ship:

File: rustler/src/types/map.rs [hunk 1]
```diff
@@ -31,12 +31,12 @@ impl<'a> Term<'a> {
     /// ```
     pub fn map_from_arrays(
         env: Env<'a>,
-        keys: &[Term<'a>],
-        values: &[Term<'a>],
+        keys: &[impl Encoder],
+        values: &[impl Encoder],
```
Comment (filmor) [inline rustler/src/types/map.rs:35]: Shouldn't this be `impl Encoder + 'a` as @hansihe noted?

Comment (SeokminHong) [reply rustler/src/types/map.rs:35]: `Encoder::encode` encodes a Rust value into a term bounds to `env`'s lifetime. So, it actually doesn't need the lifetime bound!

File: rustler/src/types/elixir_struct.rs [hunk 2]
```diff
@@ -12,17 +12,15 @@ use super::map::map_new;
...
     // In an Elixir struct the value in the __struct__ field is always an atom.
-    map.map_get(atom::__struct__().to_term(env))
-        .and_then(Atom::from_term)
+    map.map_get(atom::__struct__()).and_then(Atom::from_term)
 }
 
 pub fn make_ex_struct<'a>(env: Env<'a>, struct_module: &str) -> NifResult<Term<'a>> {
     let map = map_new(env);
 
-    let struct_atom = atom::__struct__().to_term(env);
```
Comment (evnu) [inline rustler/src/types/elixir_struct.rs:24]: The PR would work without these changes as well, correct? But we can save some work as `to_term(env)` is essentially what happens inside of `map_put()` anyways? (I wonder if the PR is breaking our current APIs or not)

File: rustler/src/types/map.rs [hunk 3]
```diff
@@ -160,27 +153,18 @@ impl<'a> Term<'a> {
...
-
-        assert!(
-            map_env == key.get_env(),
-            "key is from different environment as map"
-        );
-        assert!(
-            map_env == new_value.get_env(),
-            "value is from different environment as map"
-        );
+    pub fn map_update(self, key: impl Encoder, new_value: impl Encoder) -> NifResult<Term<'a>> {
```
[the remark this answers is not in the document]
Comment (evnu) [reply rustler/src/types/map.rs:156]: Note that there is an `impl Encoder` for term, so we could pass in two terms from different environments I think.

**Review by evnu — COMMENTED** [review]
This looks nice. I am a little confused if we need to take care of the `Env` somehow. Users might come up with the idea of passing `Term`s into the functions and have different environments for the terms.

File: rustler/src/types/map.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (SeokminHong) [reply rustler/src/types/map.rs:156]: They will be re-encoded for `env`. You know, the implementation for `Term` calls `self.in_env(env)` internally, so if the key and value are terms from different environments, they will be created as copied terms on the map's environment.

[the remark this answers is not in the document]
Comment (evnu) [reply rustler/src/types/map.rs:156]: Ah, perfect, thanks for the explanation. I missed that `Term::in_env()` copies the term if necessary.

**Review by evnu — APPROVED** [review]

**Pull request merged**